`singletons-th` fails to single shadowed variable properly

I would expect this code to compile:

```hs
{-# LANGUAGE GHC2021 #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Bug where

import Data.Singletons.Base.TH
import Prelude.Singletons

$(singletons [d|
  f :: a -> a
  f x = x
  |])

$(singletons [d|
  g :: (a -> a) -> a -> a
  g f x = f x
  |])
```

But it doesn't (with `singletons-base-3.2`):

```
Bug.hs:17:2: error: [GHC-25897]
    • Couldn't match type ‘t2’ with ‘Apply t1 t2’
      Expected: Sing (Apply t1 t2)
        Actual: Sing t2
      ‘t2’ is a rigid type variable bound by
        the type signature for:
          sG :: forall a (t1 :: a ~> a) (t2 :: a).
                Sing t1 -> Sing t2 -> Sing (Apply (Apply GSym0 t1) t2)
        at Bug.hs:(17,2)-(20,5)
    • In the second argument of ‘applySing’, namely ‘sX’
      In the expression: applySing (singFun1 @FSym0 sF) sX
      In an equation for ‘sG’:
          sG (sF :: Sing f_aklF) (sX :: Sing x_aklG)
            = applySing (singFun1 @FSym0 sF) sX
    • Relevant bindings include
        sX :: Sing t2 (bound at Bug.hs:17:2)
        sF :: Sing t1 (bound at Bug.hs:17:2)
        sG :: Sing t1 -> Sing t2 -> Sing (Apply (Apply GSym0 t1) t2)
          (bound at Bug.hs:17:2)
   |
17 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...

Bug.hs:17:2: error: [GHC-83865]
    • Couldn't match type: SLambda t1
                     with: Sing t3 -> Sing t3
      Expected: Sing t3 -> Sing (FSym0 @@ t3)
        Actual: Sing t1
    • In the second argument of ‘singFun1’, namely ‘sF’
      In the first argument of ‘applySing’, namely ‘(singFun1 @FSym0 sF)’
      In the expression: applySing (singFun1 @FSym0 sF) sX
    • Relevant bindings include
        sX :: Sing t2 (bound at Bug.hs:17:2)
        sF :: Sing t1 (bound at Bug.hs:17:2)
        sG :: Sing t1 -> Sing t2 -> Sing (Apply (Apply GSym0 t1) t2)
          (bound at Bug.hs:17:2)
   |
17 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...
```

The `-ddump-simpl` output for `sG` reveals why:

```hs
    sG ::
      (forall (t_aklH :: (~>) a_aklw a_aklw) (t_aklI :: a_aklw).
       Sing t_aklH
       -> Sing t_aklI
          -> Sing (Apply (Apply GSym0 t_aklH) t_aklI :: a_aklw) :: Type)
    sG (sF :: Sing f_aklF) (sX :: Sing x_aklG)
      = applySing (singFun1 @FSym0 sF) sX
```

Note that it is trying to use `singFun1 @FSym0 sF`, but that is referring to the promoted version of the top-level `f` function, not the local `f` variable. In order to be correct, this should instead be just `sF`.

Some other observations:

* This only affects singled code. If you replace each call to `singletons` with `promote`, then everything compiles.
* The use of two separate TH splices is critical. If you combine `f` and `g`'s definitions into a single TH splice, then the generated code will compile.

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -149,6 +149,7 @@ tests =
     , compileAndDumpStdTest "T555"
     , compileAndDumpStdTest "T559"
     , compileAndDumpStdTest "T567"
+    , compileAndDumpStdTest "T571"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/singletons-base/tests/compile-and-dump/Singletons/T571.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T571.golden
@@ -1,0 +1,66 @@
+Singletons/T571.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| f :: a -> a
+          f x = x |]
+  ======>
+    f :: a -> a
+    f x = x
+    type FSym0 :: (~>) a a
+    data FSym0 :: (~>) a a
+      where
+        FSym0KindInference :: SameKind (Apply FSym0 arg) (FSym1 arg) =>
+                              FSym0 a0123456789876543210
+    type instance Apply FSym0 a0123456789876543210 = F a0123456789876543210
+    instance SuppressUnusedWarnings FSym0 where
+      suppressUnusedWarnings = snd ((,) FSym0KindInference ())
+    type FSym1 :: a -> a
+    type family FSym1 (a0123456789876543210 :: a) :: a where
+      FSym1 a0123456789876543210 = F a0123456789876543210
+    type F :: a -> a
+    type family F (a :: a) :: a where
+      F x = x
+    sF ::
+      (forall (t :: a). Sing t -> Sing (Apply FSym0 t :: a) :: Type)
+    sF (sX :: Sing x) = sX
+    instance SingI (FSym0 :: (~>) a a) where
+      sing = singFun1 @FSym0 sF
+Singletons/T571.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| g :: (a -> a) -> a -> a
+          g f x = f x |]
+  ======>
+    g :: (a -> a) -> a -> a
+    g f x = f x
+    type GSym0 :: (~>) ((~>) a a) ((~>) a a)
+    data GSym0 :: (~>) ((~>) a a) ((~>) a a)
+      where
+        GSym0KindInference :: SameKind (Apply GSym0 arg) (GSym1 arg) =>
+                              GSym0 a0123456789876543210
+    type instance Apply GSym0 a0123456789876543210 = GSym1 a0123456789876543210
+    instance SuppressUnusedWarnings GSym0 where
+      suppressUnusedWarnings = snd ((,) GSym0KindInference ())
+    type GSym1 :: (~>) a a -> (~>) a a
+    data GSym1 (a0123456789876543210 :: (~>) a a) :: (~>) a a
+      where
+        GSym1KindInference :: SameKind (Apply (GSym1 a0123456789876543210) arg) (GSym2 a0123456789876543210 arg) =>
+                              GSym1 a0123456789876543210 a0123456789876543210
+    type instance Apply (GSym1 a0123456789876543210) a0123456789876543210 = G a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings (GSym1 a0123456789876543210) where
+      suppressUnusedWarnings = snd ((,) GSym1KindInference ())
+    type GSym2 :: (~>) a a -> a -> a
+    type family GSym2 (a0123456789876543210 :: (~>) a a) (a0123456789876543210 :: a) :: a where
+      GSym2 a0123456789876543210 a0123456789876543210 = G a0123456789876543210 a0123456789876543210
+    type G :: (~>) a a -> a -> a
+    type family G (a :: (~>) a a) (a :: a) :: a where
+      G f x = Apply f x
+    sG ::
+      (forall (t :: (~>) a a) (t :: a).
+       Sing t -> Sing t -> Sing (Apply (Apply GSym0 t) t :: a) :: Type)
+    sG (sF :: Sing f) (sX :: Sing x) = applySing sF sX
+    instance SingI (GSym0 :: (~>) ((~>) a a) ((~>) a a)) where
+      sing = singFun2 @GSym0 sG
+    instance SingI d => SingI (GSym1 (d :: (~>) a a) :: (~>) a a) where
+      sing = singFun1 @(GSym1 (d :: (~>) a a)) (sG (sing @d))
+    instance SingI1 (GSym1 :: (~>) a a -> (~>) a a) where
+      liftSing (s :: Sing (d :: (~>) a a))
+        = singFun1 @(GSym1 (d :: (~>) a a)) (sG s)

--- a/singletons-base/tests/compile-and-dump/Singletons/T571.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T571.hs
@@ -1,0 +1,13 @@
+module T571 where
+
+import Data.Singletons.TH
+
+$(singletons [d|
+  f :: a -> a
+  f x = x
+  |])
+
+$(singletons [d|
+  g :: (a -> a) -> a -> a
+  g f x = f x
+  |])

--- a/singletons-th/CHANGES.md
+++ b/singletons-th/CHANGES.md
@@ -15,6 +15,8 @@ next [????.??.??]
   ```
 * Fix a bug in which data types using visible dependent quantification would
   generate ill-scoped code when singled.
+* Fix a bug in which singling a local variable that shadows a top-level
+  definition would fail to typecheck in some circumstances.
 
 3.2 [2023.03.12]
 ----------------

--- a/singletons-th/src/Data/Singletons/TH/Names.hs
+++ b/singletons-th/src/Data/Singletons/TH/Names.hs
@@ -167,6 +167,11 @@ pureName = 'pure
 apName = '(<*>)
 liftA2Name = 'liftA2
 
+-- | Return a fresh alphanumeric 'Name'. In particular, if the supplied 'Name'
+-- is symbolic (e.g., (%%)), then return a fresh 'Name' with the 'OccName' @ty@.
+-- Otherwise, return a fresh 'Name' with the same 'OccName' as the supplied
+-- 'Name'. See @Note [Tracking local variables]@ in
+-- "Data.Singletons.TH.Promote.Monad" for why we do this.
 mkTyName :: Quasi q => Name -> q Name
 mkTyName tmName = do
   let nameStr  = nameBase tmName

--- a/singletons-th/src/Data/Singletons/TH/Promote/Monad.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Monad.hs
@@ -24,21 +24,24 @@ import Language.Haskell.TH.Desugar.OMap.Strict (OMap)
 import Data.Singletons.TH.Options
 import Data.Singletons.TH.Syntax
 
-type LetExpansions = OMap Name DType  -- from **term-level** name
-
 -- environment during promotion
 data PrEnv =
-  PrEnv { pr_options      :: Options
-        , pr_lambda_bound :: OMap Name Name
-        , pr_let_bound    :: LetExpansions
-        , pr_local_decls  :: [Dec]
+  PrEnv { pr_options     :: Options
+        , pr_lambda_vars :: OMap Name Name
+          -- ^ Map from term-level 'Name's of variables bound in lambdas and
+          -- function clauses to their type-level counterparts.
+          -- See @Note [Tracking local variables]@.
+        , pr_local_vars  :: OMap Name DType
+          -- ^ Map from term-level 'Name's of local variables to their
+          -- type-level counterparts. See @Note [Tracking local variables]@.
+        , pr_local_decls :: [Dec]
         }
 
 emptyPrEnv :: PrEnv
-emptyPrEnv = PrEnv { pr_options      = defaultOptions
-                   , pr_lambda_bound = OMap.empty
-                   , pr_let_bound    = OMap.empty
-                   , pr_local_decls  = [] }
+emptyPrEnv = PrEnv { pr_options     = defaultOptions
+                   , pr_lambda_vars = OMap.empty
+                   , pr_local_vars  = OMap.empty
+                   , pr_local_decls = [] }
 
 -- the promotion monad
 newtype PrM a = PrM (ReaderT PrEnv (WriterT [DDec] Q) a)
@@ -55,14 +58,8 @@ instance OptionsMonad PrM where
 -- return *type-level* names
 allLocals :: MonadReader PrEnv m => m [Name]
 allLocals = do
-  lambdas <- asks (OMap.assocs . pr_lambda_bound)
-  lets    <- asks pr_let_bound
-    -- filter out shadowed variables!
-  return [ typeName
-         | (termName, typeName) <- lambdas
-         , case OMap.lookup termName lets of
-             Just (DVarT typeName') | typeName' == typeName -> True
-             _                                              -> False ]
+  lambdas <- asks (OMap.assocs . pr_lambda_vars)
+  return $ map snd lambdas
 
 emitDecs :: MonadWriter [DDec] m => [DDec] -> m ()
 emitDecs = tell
@@ -72,27 +69,38 @@ emitDecsM action = do
   decs <- action
   emitDecs decs
 
--- when lambda-binding variables, we still need to add the variables
--- to the let-expansion, because of shadowing. ugh.
+-- ^ Bring a list of 'VarPromotions' into scope for the duration the supplied
+-- computation. See @Note [Tracking local variables]@.
 lambdaBind :: VarPromotions -> PrM a -> PrM a
 lambdaBind binds = local add_binds
-  where add_binds env@(PrEnv { pr_lambda_bound = lambdas
-                             , pr_let_bound    = lets }) =
-          let new_lets = OMap.fromList [ (tmN, DVarT tyN) | (tmN, tyN) <- binds ] in
-          env { pr_lambda_bound = OMap.fromList binds `OMap.union` lambdas
-              , pr_let_bound    = new_lets            `OMap.union` lets }
+  where add_binds env@(PrEnv { pr_lambda_vars = lambdas
+                             , pr_local_vars  = locals }) =
+          -- Per Note [Tracking local variables], these will be added to both
+          -- `pr_lambda_vars` and `pr_local_vars`.
+          let new_locals = OMap.fromList [ (tmN, DVarT tyN) | (tmN, tyN) <- binds ] in
+          env { pr_lambda_vars = OMap.fromList binds `OMap.union` lambdas
+              , pr_local_vars  = new_locals          `OMap.union` locals }
 
+-- ^ A pair consisting of a term-level 'Name' of a variable, bound in a @let@
+-- binding or @where@ clause, and its type-level counterpart.
+-- See @Note [Tracking local variables]@.
 type LetBind = (Name, DType)
+
+-- ^ Bring a list of 'LetBind's into scope for the duration the supplied
+-- computation. See @Note [Tracking local variables]@.
 letBind :: [LetBind] -> PrM a -> PrM a
 letBind binds = local add_binds
-  where add_binds env@(PrEnv { pr_let_bound = lets }) =
-          env { pr_let_bound = OMap.fromList binds `OMap.union` lets }
+  where add_binds env@(PrEnv { pr_local_vars = locals }) =
+          env { pr_local_vars = OMap.fromList binds `OMap.union` locals }
 
+-- | Map a term-level 'Name' to its type-level counterpart. This function is
+-- aware of any local variables that are currently in scope.
+-- See @Note [Tracking local variables]@.
 lookupVarE :: Name -> PrM DType
 lookupVarE n = do
   opts <- getOptions
-  lets <- asks pr_let_bound
-  case OMap.lookup n lets of
+  locals <- asks pr_local_vars
+  case OMap.lookup n locals of
     Just ty -> return ty
     Nothing -> return $ DConT $ defunctionalizedName0 opts n
 
@@ -115,3 +123,123 @@ promoteMDecs :: OptionsMonad q => [Dec] -> PrM [DDec] -> q [DDec]
 promoteMDecs locals thing = do
   (decs1, decs2) <- promoteM locals thing
   return $ decs1 ++ decs2
+
+{-
+Note [Tracking local variables]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Handling local variables in singletons-th requires some care. There are two
+sorts of local variables that singletons-th tracks:
+
+1. Lambda-bound variables, e.g.,
+
+     f = \x -> x
+     g x = x
+
+   In both `f` and `g`, the variable `x` is considered lambda-bound.
+
+2. Let-bound variables, e.g.,
+
+     h =
+       let x = 42 in
+       x + x
+
+     i = x + x
+       where
+         x = 42
+
+   In both `h` and `i`, the variable `x` is considered let-bound.
+
+Why does singletons-th need to track local variables? It's because they must
+be promoted differently depending on whether they are local or not. Consider:
+
+  j = ... x ...
+
+When promoting the `j` function to a type family `J`, there are three possible
+ways of promoting `x`:
+
+* If `x` is a lambda-bound variable, then `x` must be promoted to a type
+  variable. In general, we cannot promote `x` to the same name. Consider this
+  example:
+
+    j (%%) x y = x %% y
+
+  Here, `(%%)`, `x`, and `y` are lambda-bound variables. But we cannot promote
+  `j` to this type family:
+
+    type family J (%%) x y where
+      J (%%) x y = x %% y
+
+  This is because type variable names cannot be symbolic like `(%%)` is. As a
+  result, we create a fresh name `ty` and promote each occurrence of `(%%)` to
+  `ty`:
+
+    type family J ty x y where
+      J ty x y = x `ty` y
+
+  See `mkTyName` in Data.Singletons.TH.Names. In fact, `mkTyName` will also
+  freshen alphanumeric names, so it would be more accurate to say that `j` will
+  be promoted to this:
+
+    type family J ty x_123 y_456 where
+      J ty x_123 y_456 = x_123 `ty` y_456
+
+  Where `x_123` and `y_456` are fresh names that are distinct from `x` and `y`.
+  Freshening alphanumeric names like `x` and `y` is probably not strictly
+  necessary, but `mkTyName` does it anyway (1) for consistency with symbolic
+  names and (2) to make the type-level names easier to tell apart from the
+  original term-level names.
+
+* If `x` is a let-bound variable, then `x` must be promoted to something like
+  `LetX`, where `LetX` is the lambda-lifted version of `x`. For instance, we
+  would promote this:
+
+    j = x
+      where
+        x = True
+
+  To this:
+
+    type family J where
+      J = LetX
+    type family LetX where
+      LetX = True
+
+* If `x` is not a local variable at all, then `x` must be promoted to something
+  like `X`, which is assumed to be a top-level function. For instance, we would
+  promote this:
+
+    x = 42
+    j = x
+
+  To this:
+
+    type family X where
+      X = 42
+    type family J where
+      J = X
+
+Being able to distinguish between all these sorts of variables requires
+recording whether they are lambda-/let-bound at their binding sites during
+promotion and singling. During promotion, the `pr_local_vars` field of `PrEnv`
+is responsible for this. In singling, the `sg_local_vars` field of `SgEnv` is
+responsible for this. Each of these fields are Maps from the original,
+term-level Names to the promoted or singled versions of the Names. The
+`lookupVarE` functions (which can be found in both
+Data.Singletons.TH.Promote.Monad and Data.Singletons.TH.Single.Monad) are
+responsible for determining what a term-level Name should be mapped to.
+
+In addition to `pr_local_vars` and `sg_local_vars`, which include both lambda-
+and let-bound variables, `PrEnv` also includes a separate `pr_lambda_vars`
+field, which only tracks lambda-bound variables. We must do this because
+lambda-bound variables are treated differently during lambda lifting.
+Lambda-lifted functions must close over any lambda-bound variables in scope,
+but /not/ any let-bound variables in scope, since the latter are lambda-lifted
+separately. (Because singling does not do anything akin to lambda lifting,
+`SgEnv` does not have anything like `sg_lambda_vars`.)
+
+A consequence of this is that when we lambda-bind a variable during promotion
+(see `lambdaBind`), we add the variable to both `pr_lambda_vars` and
+`pr_local_vars`.  When we let-bind a variable during promotion (see `letBind`),
+we only add the variable to `pr_local_vars`. This means that `pr_lambda_vars`
+will always be a subset of `pr_local_vars`.
+-}

--- a/singletons-th/src/Data/Singletons/TH/Single.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single.hs
@@ -716,8 +716,10 @@ singLetDecRHS cxts name ld_rhs = do
 
 singClause :: ADClause -> SgM DClause
 singClause (ADClause var_proms pats exp) = do
+  opts <- getOptions
   (sPats, sigPaExpsSigs) <- evalForPair $ mapM (singPat (Map.fromList var_proms)) pats
-  sBody <- singExp exp
+  let lambda_binds = map (\(n,_) -> (n, singledValueName opts n)) var_proms
+  sBody <- bindLambdas lambda_binds $ singExp exp
   return $ DClause sPats $ mkSigPaCaseE sigPaExpsSigs sBody
 
 singPat :: Map Name Name   -- from term-level names to type-level names
@@ -893,7 +895,7 @@ singExp (ADAppE e1 e2) = do
 singExp (ADLamE ty_names prom_lam names exp) = do
   opts <- getOptions
   let sNames = map (singledValueName opts) names
-  exp' <- singExp exp
+  exp' <- bindLambdas (zip names sNames) $ singExp exp
   -- we need to bind the type variables... but DLamE doesn't allow SigT patterns.
   -- So: build a case
   let caseExp = DCaseE (mkTupleDExp (map DVarE sNames))
@@ -995,8 +997,10 @@ isException (DStaticE e)          = isException e
 
 singMatch :: ADMatch -> SgM DMatch
 singMatch (ADMatch var_proms pat exp) = do
+  opts <- getOptions
   (sPat, sigPaExpsSigs) <- evalForPair $ singPat (Map.fromList var_proms) pat
-  sExp <- singExp exp
+  let lambda_binds = map (\(n,_) -> (n, singledValueName opts n)) var_proms
+  sExp <- bindLambdas lambda_binds $ singExp exp
   return $ DMatch sPat $ mkSigPaCaseE sigPaExpsSigs sExp
 
 singLit :: Lit -> SgM DExp

--- a/singletons-th/src/Data/Singletons/TH/Single/Monad.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single/Monad.hs
@@ -11,7 +11,8 @@ The SgM monad allows reading from a SgEnv environment and is wrapped around a Q.
 -}
 
 module Data.Singletons.TH.Single.Monad (
-  SgM, bindLets, bindContext, askContext, lookupVarE, lookupConE,
+  SgM, bindLambdas, bindLets, bindContext,
+  askContext, lookupVarE, lookupConE,
   wrapSingFun,
   singM, singDecsM,
   emitDecs, emitDecsM
@@ -35,14 +36,17 @@ import Control.Applicative
 -- environment during singling
 data SgEnv =
   SgEnv { sg_options     :: Options
-        , sg_let_binds   :: Map Name DExp   -- from the *original* name
+        , sg_local_vars  :: Map Name DExp
+          -- ^ Map from term-level 'Name's of local variables to their
+          -- singled counterparts. See @Note [Tracking local variables]@ in
+          -- "Data.Singletons.TH.Promote.Monad".
         , sg_context     :: DCxt -- See Note [Tracking the current type signature context]
         , sg_local_decls :: [Dec]
         }
 
 emptySgEnv :: SgEnv
 emptySgEnv = SgEnv { sg_options     = defaultOptions
-                   , sg_let_binds   = Map.empty
+                   , sg_local_vars  = Map.empty
                    , sg_context     = []
                    , sg_local_decls = []
                    }
@@ -59,10 +63,24 @@ instance DsMonad SgM where
 instance OptionsMonad SgM where
   getOptions = asks sg_options
 
+-- ^ Bring a list of lambda-bound names into scope for the duration the supplied
+-- computation, where the first element of each pair is the original, term-level
+-- name, and the second element of each pair is the singled counterpart.
+-- See @Note [Tracking local variables]@ in "Data.Singletons.TH.Promote.Monad".
+bindLambdas :: [(Name, Name)] -> SgM a -> SgM a
+bindLambdas lambdas = local add_binds
+  where add_binds env@(SgEnv { sg_local_vars = locals }) =
+          let new_locals = Map.fromList [ (tmN, DVarE tyN) | (tmN, tyN) <- lambdas ] in
+          env { sg_local_vars = new_locals `Map.union` locals }
+
+-- ^ Bring a list of let-bound names into scope for the duration the supplied
+-- computation, where the first element of each pair is the original, term-level
+-- name, and the second element of each pair is the singled counterpart.
+-- See @Note [Tracking local variables]@ in "Data.Singletons.TH.Promote.Monad".
 bindLets :: [(Name, DExp)] -> SgM a -> SgM a
-bindLets lets1 =
-  local (\env@(SgEnv { sg_let_binds = lets2 }) ->
-               env { sg_let_binds = (Map.fromList lets1) `Map.union` lets2 })
+bindLets lets =
+  local (\env@(SgEnv { sg_local_vars = locals }) ->
+               env { sg_local_vars = Map.fromList lets `Map.union` locals })
 
 -- Add some constraints to the current type signature context.
 -- See Note [Tracking the current type signature context]
@@ -76,12 +94,16 @@ bindContext ctxt1
 askContext :: SgM DCxt
 askContext = asks sg_context
 
+-- | Map a term-level 'Name' to its singled counterpart. This function is aware
+-- of any local variables that are currently in scope.
+-- See @Note [Tracking local variables]@ in "Data.Singletons.TH.Promote.Monad".
 lookupVarE :: Name -> SgM DExp
 lookupVarE name = do
   opts <- getOptions
   lookup_var_con (singledValueName opts)
                  (DVarE . singledValueName opts) name
 
+-- | Map a data constructor name to its singled counterpart.
 lookupConE :: Name -> SgM DExp
 lookupConE name = do
   opts <- getOptions
@@ -91,9 +113,9 @@ lookupConE name = do
 lookup_var_con :: (Name -> Name) -> (Name -> DExp) -> Name -> SgM DExp
 lookup_var_con mk_sing_name mk_exp name = do
   opts <- getOptions
-  letExpansions <- asks sg_let_binds
+  localExpansions <- asks sg_local_vars
   sName <- mkDataName (nameBase (mk_sing_name name)) -- we want *term* names!
-  case Map.lookup name letExpansions of
+  case Map.lookup name localExpansions of
     Nothing -> do
       -- try to get it from the global context
       m_dinfo <- liftM2 (<|>) (dsReify sName) (dsReify name)
@@ -138,7 +160,8 @@ singDecsM locals thing = do
 {-
 Note [Tracking the current type signature context]
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-Much like we track the let-bound names in scope, we also track the current
+Much like we track the locally-bound names in scope (see Note [Tracking local
+variables] in Data.Singletons.TH.Promote.Monad), we also track the current
 context. For instance, in the following program:
 
   -- (1)

